crates/router/src/core/payments/operations/payment_confirm.rs

```diff
@@ -479,3 +499,99 @@ impl<F: Send + Clone> ValidateRequest<F, api::PaymentsRequest> for PaymentConfir
+    request: &api::PaymentsRequest,
+) -> RouterResult<()> {
+    let parsed_metadata_db: Option<api_models::payments::Metadata> = payment_intent
+        .clone()
+        .metadata
+        .map(|metadata_value| {
+            metadata_value
+                .peek()
+                .clone()
+                .parse_value("metadata")
+                .change_context(errors::ApiErrorResponse::InvalidDataValue {
+                    field_name: "metadata",
+                })
+                .attach_printable("unable to parse metadata")
+        })
+        .transpose()
+        .unwrap_or_default();
+    let order_details_metadata_db = parsed_metadata_db
+        .clone()
+        .and_then(|meta| meta.order_details);
+    let order_details_outside_metadata_db = payment_intent.clone().order_details;
+    let order_details_outside_metadata_req = request.clone().order_details;
+    let order_details_metadata_req = request.clone().metadata.and_then(|meta| meta.order_details);
+
+    if order_details_outside_metadata_req
+        .clone()
+        .zip(order_details_metadata_req.clone())
+        .is_some()
+    {
+        Err(errors::ApiErrorResponse::NotSupported { message: "order_details cannot be present both inside and outside metadata in payments request".to_string() })?
+    }
+    if order_details_metadata_db
+        .clone()
+        .zip(order_details_outside_metadata_db.clone())
+        .is_some()
+    {
+        Err(errors::ApiErrorResponse::NotSupported { message: "order_details cannot be present both inside and outside metadata in payment intent in db".to_string() })?
+    }
+    let order_details_outside = match order_details_outside_metadata_req {
+        Some(order) => match order_details_metadata_db {
+            Some(_) => Err(errors::ApiErrorResponse::NotSupported {
+                message: "order_details previously present inside of metadata".to_string(),
+            })?,
+            None => Some(order),
+        },
+        None => match order_details_metadata_req {
+            Some(_order) => match order_details_outside_metadata_db {
+                Some(_) => Err(errors::ApiErrorResponse::NotSupported {
+                    message: "order_details previously present outside of metadata".to_string(),
+                })?,
+                None => None,
+            },
+            None => None,
+        },
+    };
+    let metadata_with_order_details = request.clone().metadata.map(|meta| {
+        let transformed_metadata = match parsed_metadata_db {
+            Some(meta_db) => api_models::payments::Metadata {
+                order_details: meta.order_details,
+                ..meta_db
+            },
+            None => meta,
+        };
+        let transformed_metadata_value =
+            Encode::<api_models::payments::Metadata>::encode_to_value(&transformed_metadata)
+                .change_context(errors::ApiErrorResponse::InternalServerError)
```

Comment:
Is this a InternalServerError?

- When this encoding of data fails?
- If the value is validated before means this won't happen itself?

Reply:
we are fetching metadata from db, and adding our order_details field to it (metadata_with_order_details).
Now this is a struct of type Metadata.
In the above code,  we are trying to encode Metadata to a value...it should never fail , as metadata struct is already type checked by rust during compilation.
but if it fails, its not a problem of the payments request. Thus it should be an internal server error.